Additions to wiki - Model Interpretation Documentation (#75)
ccao-data/wiki: SOPs/Model-Evaluation-ML-metrics.md

Reviewer: I would replace this outline with a "terms" section and then try to clean up the constant switching between population, sample, sales, and assessment. It's a lot of parentheses and extra language that we could get out of the way super quick and then use a couple small, consistent terms throughout. I tried to clean this up in the word doc but perhaps I made it worse.

> Useful terms
> - Sample: the universe of parcel sales we use to train and test our model
> - Population: the universe of parcels that the model needs to value

etc...

Author: Great call.  I think this is actually really important to be clear on- (population, sample, sales, and assessment.) - as it can be a source of confusion when discussing different model outputs and types of evaluation (especially with regard to differences between ml evaluation and domain specific evaluation).

# Model Validation, Interpretation, and Testing Guidelines

After running the model you will need to interpret the results. First, assess how well the sample data (sales sample, training data) matches the population (assessment set, all parcels to be assessed), then note any factors that may be impacting your model performance such as data issues or novel market trends. Finally, interpret the formal model performance statistics to select the most accurate and generalizable model, that conforms to IAAO standards.

> **Overview:**
> 
> 1. Assessing how representative your sales sample is of the assessment set.
>    - a. Balance tests
>    - b. Visual inspection
>    - c. Not missing at random
>    - d. Domain specific approach
> 
> 2. Noting any real-world housing market changes that may impact your model, and/or interactions between data and model that may affect your results (model drift, data drift).
> 
> 3. Interpreting model performance (evaluating machine learning and assessment metrics).
> 

Reviewer: ```suggestion

```

There is an "Outline" button next to markdown files that already provides this feature in a really clean way:

<img width="612" height="711" alt="Image" src="https://github.com/user-attachments/assets/0248ae8e-5a4c-4361-9231-41c7787feb39" />

Or, if we're committed to this outline, i'd link to the sections through it using [section links](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#section-links).

---

## 1. Assessing How Representative Your Sales Sample Is of the Assessment Set

> To ensure that a model can generalize from a sample to the population, we need to check that our sales sample contains buildings that are similar to those in the entire population. We should generally see parcels with the same composition of features, in the same proportions, in the sales (sample) and assessment set (population). We check for this with statistical tests and visual inspections of distributions.

Reviewer: ```suggestion
To ensure that a model is generalizable, we need to check that our sales sample is similar to the population. We should see parcels with the same composition of features, in the same proportions, in the sales (sample) and assessment set (population). We check for this with statistical tests and visual inspections of distributions.
```

To ensure that a model can generalize from a sample (the sales data that we train the model on) to the population (all buildings being assessed), we need to check that our sales sample contains buildings that are similar to those in the entire population:  That is, we should generally see buildings with the same composition of features, in the same proportions, in the sales-sample and in the population.  We check for this in several ways (primarily with statistical tests and visual inspections of distributions).

*(Note: if there are factors that cause some properties (and their attendant sale-prices) to be over-represented in the sample, the model will over-index to these types of properties, likely leading to over or under-valuation).*

---

Ways that we test for differences between the sample and the population (and some possible corrections):  In a perfectly matched sample, no feature would predict inclusion/exclusion of a property in the sales-sample. That is, all properties, of all types would have an equal chance of being sold in a given year.   These tests allow us to test that assumption (and to develop possible corrections).

### A. Balance Tests

> *(See the "Statistical Tests" section of the model performance report.)*  In a perfectly matched sample, no feature would predict inclusion/exclusion of a property in the sales-sample. Any feature that predicts inclusion in the sales set at a level greater than chance (statistical significance) suggests that this feature is over-or under-represented in the sample and will likely bias your results.  (This is especially the case for features that also turn out to have high shap values in your results).  To check this, we run a simple logistic regression predicting the likelihood-of-a-sale, given a property's features. The resulting p values (for each feature in the report) tells you that a feature predicts inclusion in the sample at a level greater than expected-due-to-chance, while the Beta value gives you a relative sense of the weight (importance) and direction (include vs exclude) of that feature.  (In our report, asterisks, represent statistically significant predictors).  (Low p-values suggest statistical significance, high magnitudes for the Betas suggest a large impact).  When a feature is predictive of inclusion in the sample, this means that your sample is likely biased towards properties with this feature, and may thus value these, or other properties inaccurately.

Reviewer 2: I also like the note of where we can find this. It may be a bit too nitty to do this for every section, but maybe under the big headers, note which sections of reports we can find the different interpretations.

- **c.** To validate your investigations from the balance tests, look at the standardized mean differences between the sales sample and assessment set for each feature. Larger differences indicate a more likely deviation between the sales sample and assessment set.

#### II. Visual Inspection

See empirical distributions on the performance report. The distribution of a feature in the sales sample should visually match those in the assessment set. We only calculate this for the full sales sample, but it may differ at the township or neighborhood-level. We could apply a KS test to check if the feature distributions between the sales sample and assessment set are the same.

> #### III. Non-Missing at Random

Reviewer 3: Possible typo for "Missing Not At Random" ([MNAR](https://aml4td.org/chapters/missing-data.html))? It's been a minute since I've dusted off the textbooks.

``` suggestion
#### III. Missing Not at Random
```

Author: yep- sorry, typo.

> "Non-missing at random" means that some feature or a particular value of a feature is missing in a way that's correlated with the outcome variable or another variable in the dataset. This can indicate systemic undersampling. In our case, it is somewhat controlled for by the fact that lgbm actually incorporates missing values as a predictor\*. Because of this we currently don't track correlations of nulls as rigorously as we otherwise might, though you can get a sense of the percentage of missingness for each feature by looking at the "Missingness" heading in the Feature report.

Reviewer 3: See above; as defined, this sounds to me like MNAR (missing not at random).

```suggestion
"Missing not at random" means that some feature or a particular value of a feature is missing in a way that's correlated with the outcome variable or another variable in the dataset. This can indicate systemic undersampling. In our case, it is somewhat controlled for by the fact that lgbm actually incorporates missing values as a predictor\*. Because of this we currently don't track correlations of nulls as rigorously as we otherwise might, though you can get a sense of the percentage of missingness for each feature by looking at the "Missingness" heading in the Feature report.
```

> > \* See: [How does LGBM deal with missing values?](https://www.google.com/search?q=how+does+lgbm+deal+with+missing+values) | [How do XGBoost, LightGBM, and CatBoost Handle Missing Features?](https://www.google.com/search?q=how+do+xgboost+lightgbm+catboost+handle+missing+features)

Reviewer 3: Hmm, I'm wary of trusting that google's search results are mutable across people's devices. Could you link a specific source you'd recommend? Or, barring that, the LGBM [docs](https://lightgbm.readthedocs.io/en/latest/Advanced-Topics.html)?

Author: sorry- had a proper link for that- copied the wrong thing.  will update.

Author: links = fixed.

#### IV. Domain Specific Sanity Check

> Compare year-over-year changes in assessed values for sold-and-unsold houses. This is documented in the performance report under "Change In and Out of Sample". The sold and unsold properties should have roughly similar changes in assessed values.

Reviewer 3: Hm, in theory this is true... in practice, I'm not sure it always works out this neatly. Consider the case where the sold properties are all new builds in a neighborhood that is otherwise comprised of older homes. Because the sold properties are different from the rest of the population of unsold properties, might it be proper for the model to have different predictions? 

This might also reasonably happen if, hypothetically, the recently sold properties had prior assessment changes that their unsold neighbors did not.

What do you think about just softening the language a bit to something like:

```suggestion
Compare year-over-year changes in assessed values for sold and unsold houses. This is documented in the performance report under "Change In and Out of Sample". 

The sold and unsold properties should have roughly similar changes in assessed values, on the assumption that sold and unsold properties have similar characteristics and assessment histories.
```

> **Note:** Since we use a boosted model with historic data and retrain each year, we are less subject to major problems with data or model drift. However, this can still pose an issue if we have high temporal volatility and low recent sales volume.

See: https://www.ibm.com/think/topics/model-drift

---

## 3. Interpreting Model Performance Statistics

> We calculate traditional machine learning metrics and assessment-specific metrics to assess the model. The machine learning metrics we calculate are RMSE, mDape, and R-squared. The assessment metrics that we calculate and attend to are Median Ratio and Coefficient of Dispersion (COD) for accuracy and precision (respectively). We supplement our analysis with measures of vertical equity (how equal are assessments at all price levels). These are PRB, PRD, MKI, and ratio curves. These assessment metrics are explained in detail in [Mass Appraisal For The Masses: The Basics by Lars Doucet](https://www.google.com/search?q=Mass+Appraisal+For+The+Masses+The+Basics+Lars+Doucet). MKI is outlined in depth in [A Gini measure for vertical equity in property assessments](https://www.google.com/search?q=gini+measure+vertical+equity+property+assessments). See also our guide to ratio studies: https://github.com/ccao-data/wiki/blob/master/SOPs/Sales-Ratio-Studies.md

Reviewer 3: [nitpick, optional] Suggested link cleanup. Also I think we typically abbreviate median abs percentage error as MdAPE.

```suggestion
We calculate traditional machine learning metrics and assessment-specific metrics to assess the model. The machine learning metrics we calculate are RMSE, MdAPE, and R-squared. The assessment metrics that we calculate and attend to are Median Ratio and Coefficient of Dispersion (COD) for accuracy and precision (respectively). We supplement our analysis with measures of vertical equity (how equal are assessments at all price levels). These are PRB, PRD, MKI, and ratio curves. These assessment metrics are explained in detail in [Mass Appraisal For The Masses: The Basics by Lars Doucet](https://progressandpoverty.substack.com/p/mass-appraisal-for-the-masses-the). MKI is outlined in depth in [A Gini measure for vertical equity in property assessments](https://doi.org/10.63642/1357-1419.1225). See also our guide to ratio studies: https://github.com/ccao-data/wiki/blob/master/SOPs/Sales-Ratio-Studies.md
```

> Our approach is to fit candidate models using the machine learning (ML) metrics then compare/evaluate our candidate models based on both the ML metrics and the assessment metrics. We can think of the assessment metrics as acceptable bounds for candidate models and as a way to "break ties" between models that are otherwise similar in terms of ML metrics. For example, if a model shows good ML fits but is outside the acceptable assessment bounds, we might ignore it.

Reviewer 3: I think it's good to distinguish the formal fitting based on objective function, from how we choose our final model from among candidate models. While we do fit our candidate models on RMSE, we typically rely on assessment metrics like median ratio, COD, and vertical equity to actually choose our final model. 

Historically, the performant models that perform well on the assessment metrics also have lower RMSE, so I'm proposing condensing some of the text.

What do you think about this?

```suggestion
Our approach while fitting candidate models is to follow ML best practices. During model training and fitting, we use standard machine learning (ML) metrics, like RMSE (described below).

To compare and evaluate our candidate models, however, we use both ML metrics and assessment metrics. We rely heavily on median ratio, COD, and vertical equity in our recommendation for what should be the final model. 
```

Further discussion [here](https://stats.stackexchange.com/questions/242787/how-to-interpret-root-mean-squared-error-rmse-vs-standard-deviation) — more formally, here: [Shmueli, G., Bruce, P. C., Stephens, M., & Patel, N. R. (2016). *Data Mining for Business Analytics: Concepts, Techniques, and Applications with JMP Pro* (3rd Edition). Wiley.](https://www.amazon.fr/Data-Mining-Business-Analytics-Applications/dp/1118877438/)

> #### mDape (Median Absolute Percentage Error)

Reviewer 3: [nitpick, optional] I usually see this abbreviated differently:

```suggestion
MdAPE
```

> [Zillow's performance metric of choice](https://stats.stackexchange.com/questions/596324/is-median-absolute-percentage-error-useless#:~:text=Percentage%20Error%20Asymmetry:%20A%20significant%20drawback%20of,same%20factor%20yields%20only%20a%2050%25%20error). Since mDape is a [median-based metric](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error), it is more robust to outliers than other measures and complements RMSE. We shouldn't use mDape as an optimization metric as it is not a proper scoring rule, and treats over-forecasts and underforecasts asymmetrically, but it is useful for comparing model performance in a manner that is more robust to outliers. For example, in a case where two models differ slightly in their RMSE, we may accept a model with a slightly higher RMSE if it has a lower mDape than its competitor. This arrangement would likely signal that the other "low RMSE" model might simply be fitting toward some high value outliers at the expense of the median property.

Reviewer 3: I haven't thought this much about MdAPE -- nice to formally document why it's not an optimization metric. Some copy edits to slim this down below. I also swapped out the wikipedia link; confusingly, it linked to mean absolute percent error. If that was intentional, can you unpack that?

```suggestion
Median absolute percentage error (MdAPE) is a [median-based metric](https://support.numxl.com/hc/en-us/articles/115001223503-MdAPE-Median-Absolute-Percentage-Error). It is more robust to outliers than other measures and complements RMSE. We shouldn't use MdAPE as an optimization metric as it is not a proper scoring rule, and treats over-forecasts and under-forecasts asymmetrically, but it is useful for comparing model performance in a manner that is more robust to outliers. For example, in a case where two models differ slightly in RMSE, we may accept a model with a slightly higher RMSE if it has a lower MdAPE. This arrangement would likely signal that the other "low RMSE" model might simply be fitting toward some high value outliers at the expense of the median property.
```

- R-squared does not necessarily measure goodness-of-fit.
- R-squared does not necessarily measure predictive error.
- R-squared does not measure how one variable explains another (it's not causal).

See [Shalizi notes](https://www.stat.cmu.edu/~cshalizi/mreg/15/lectures/10/lecture-10.pdf)

For a more sensible interpretation of goodness-of-fit, simply eyeball the table in the model performance report Estimate vs Actual (Individual Obs.). Look for changes between model estimates and actuals.

### Train/Test Splits

> We follow a standard train/test split to test our models for overfitting. Overfitting is when our model picks up spurious correlations in our sample and uses those to predict sales. In other words, our model studies the sample so deeply it loses the ability to imagine a sample that looks different. It will do a good job predicting on the sample it was trained on but will perform worse on observations it has never seen. The ability of a model to handle situations it hasn't seen is called its "generalizability".

Reviewer 3: [minor nitpick] Suggested text tweaks

```suggestion
We follow a standard [train/test split](https://github.com/ccao-data/model-res-avm/#using-training_data) to test our models for overfitting. Overfitting occurs when a model picks up spurious correlations in the training sample and uses those to predict sales. In other words, overfitting occurs when a model studies the training sample so deeply it loses the ability to generalize predictions to properties that are different from the training set. An overfit model will do a good job predicting on the sample it was trained on but will perform worse on observations it has never seen. The ability of a model to handle situations it hasn't seen is called its "generalizability". A good model is one that can generalize its predictions, such that it reasonably accurately predicts the sale prices of properties it has not been trained on.
```

Further reading — Bias Variance Trade-off:
- https://en.wikipedia.org/wiki/Bias%E2%80%93variance_tradeoff
- https://www.ibm.com/think/topics/bias-variance-tradeoff

> **Finally:** If your sample is not a good match for your population, good train-test splits will only take you so far. This is because your sample lacks representative training data. This is why balance tests (see earlier section) are important.

### Assessment Metrics

Interpretations and acceptable ranges for assessment metrics can be found here:
https://github.com/ccao-data/wiki/blob/master/SOPs/Sales-Ratio-Studies.md

Longer descriptions here: [Mass Appraisal For The Masses: The Basics — by Lars Doucet](https://www.google.com/search?q=Mass+Appraisal+For+The+Masses+The+Basics+Lars+Doucet)

---

> ## Practical Process

Reviewer 3: [optional] What do you think about linking out to our annual model checklist? The below is slightly duplicative of it but I think your goal here is about specifically how to choose a final model from candidates.

```suggestion
## Practical Process

Our [annual model checklist](https://github.com/ccao-data/model-res-avm/blob/master/.github/ISSUE_TEMPLATE/annual-model-checklist.md) details the technical steps necessary to run candidate models each year. Below is an overview of how to reason statistically about model candidates to choose a final one.
```

Author: Yep, I think that makes sense link back to the model checklist!

(context/my thinking - since the curre model checklist is more of a  "how-to" in terms of data ingestion and mechanics, and less specific on the actual model runs.  Also planning to link out (from the model template) to this wiki, for advice on how run + interpret the model - (so as not to clutter the checklist).

I imagine we could create a new heading (model run + interpretation), after the "data ingestion/internal data requests/condo data requests*" in the model template, and link back to this.

* Somewhere around here:
<img width="860" height="310" alt="image" src="https://github.com/user-attachments/assets/a343c11f-db25-4551-a2bb-5ca048805f83" />
